### ISSUE TITLE
mbedtls_platform_gmtime_r windows bug

### ISSUE THREAD

**Issue by miroslav-mastny-regosys** [Mbed-TLS/mbedtls#4015]
### Description
- Type: Bug 
- Priority: Major

---------------------------------------------------------------
**OS**  
windows (C++ Builder)

**mbed TLS build:**  
Version: 2.16.9

WIN32 function gmtime_s has parameters and return value wrong order. 

Should be:
#if defined(_WIN32) && !defined(EFIX64) && !defined(EFI32)
    **return( ( gmtime_s( tt, tm_buf ) == 0 ) ? NULL: tm_buf );**
#elif !defined(PLATFORM_UTIL_USE_GMTIME)

Currently implemented:
#if defined(_WIN32) && !defined(EFIX64) && !defined(EFI32)
    **return( ( gmtime_s( tm_buf, tt ) == 0 ) ? tm_buf : NULL );**
#elif !defined(PLATFORM_UTIL_USE_GMTIME)

**Comment by gabor-mezei-arm** [comment]
It looks like the C++ Builder using the standard C function instead of the windows' one.

Can you help me about which macro is defined by C++ Builder and what are their value?
These are the interested macros:
_POSIX_VERSION
_POSIX_THREAD_SAFE_FUNCTIONS
__STDC__

**Comment by miroslav-mastny-regosys** [comment]
Any POSIG macro is not set.
Macro __STDC__ is set to 1 when building as 32bit C project and ANSI conformance compiler flag is set (it seems is not set by default and there is not direct switch in project setting - only user compiler switch field is useable for that)

C++Builder definition of gmtime_s in file <time.h>:
struct tm * _RTLENTRY _EXPFUNC gmtime_s(const time_t * _RESTRICT clock,
					struct tm * _RESTRICT result);

**Comment by gabor-mezei-arm** [comment]
The gmtime handling is not prepared for standard-C-like `gmtime_s` function in windows.

**Comment by gstrauss** [comment]
Is this still an issue?  The usage of `gmtime_s()` is centralized to a single use in `library/platform_util.c:mbedtls_platform_gmtime_r()` and the argument order looks correct to me.

**Comment by gabor-mezei-arm** [comment]
Yes, the `gmtime_s` function should be called the same way as `gmtime_r` if the `__STDC__` and `_WIN32` macros are defined.

The standard (C11):
`struct tm *gmtime_s( const time_t *restrict timer, struct tm *restrict buf );`

The MSVC provides:
`errno_t gmtime_s( struct tm* tmDest, const __time_t* sourceTime );`

**Comment by gstrauss** [comment]
ick.  Now I see that the signature is incompatible depending on the "standards" view in the headers.
https://en.cppreference.com/w/c/chrono/gmtime
Since `gmtime_s()` was added in C11 Annex K, is the definition properly hidden by checking for the following?
```
#define __STDC_WANT_LIB_EXT1__ 1
#include <time.h>
...
#ifdef __STDC_LIB_EXT1__
/* gmtime_s() C11 prototype view */
#else
/* gmtime_s() MSVC prototype view */
#endif
```

**Comment by daverodgman** [comment]
Fixed by #4211
